File: common/C++/Process/ShadowMap.cpp
//// ce script fournit des fonctions utilitaires pour les programmes
//// du livre Synthèse d'images à l'aide d'OpenGL

// consulter http://blog.tojicode.com/2012/07/using-webgldepthtexture.html

#include <iostream>
#include <fstream>
#include <sstream>
#include <stdexcept>
#include <string.h>
#include <stdlib.h>
#include <math.h>

#include <ShadowMap.h>


/**
 * constructeur de la classe ShadowMap
 * @param shadowmapsize : largeur et hauteur de la carte d'ombre
 * @param offsetfill : true si la shadow map doit activer le décalage de polygones
 * @param cullface : mettre GL_FRONT ou GL_BACK pour éliminer les faces avant ou de dos, GL_NONE pour ne rien configurer
 */
ShadowMap::ShadowMap(int shadowmapsize, bool offsetfill, GLenum cullface):
    // c'est un FBO ne contenant qu'une depth map de type texture
    FrameBufferObject(shadowmapsize, shadowmapsize, GL_NONE, GL_TEXTURE_2D, 0, GL_LINEAR)
{
    // options de réduction de l'acné
    m_OffsetFill = offsetfill;
    m_CullFace   = cullface;
    m_CullFacePrec = 0;
}


/**
 * supprime cette shadowmap
 */
ShadowMap::~ShadowMap()
{
}


/**
 * cette méthode change les modes de réduction de l'acné de surface
 * @param offsetfill : true si la shadow map doit activer le décalage de polygones
 * @param cullface : mettre GL_FRONT ou GL_BACK pour éliminer les faces avant ou de dos, GL_NONE pour ne rien configurer
 */
void ShadowMap::setAcneReduction(bool offsetfill, GLenum cullface)
{
    m_OffsetFill = offsetfill;
    m_CullFace   = cullface;
}


/**
 * redirige tous les tracés suivants vers la ShadowMap
 */
void ShadowMap::enable()
{
    // activer le FBO
    FrameBufferObject::enable();

    // (optionnel) éliminer les faces avant, afin d'éviter l'acné de surface
    glGetIntegerv(GL_CULL_FACE_MODE, &m_CullFacePrec);
    if (m_CullFace != GL_NONE) {
        glEnable(GL_CULL_FACE);
        glCullFace(m_CullFace);
    }

    // (optionnel) décalage de polygones, afin d'éviter l'acné de surface
    if (m_OffsetFill) {
        glEnable(GL_POLYGON_OFFSET_FILL);
        glPolygonOffset(1.0, 1.0);
    }

    // effacer le depth buffer
    glClear(GL_DEPTH_BUFFER_BIT);
}


/**
 * cesser de rediriger les dessins dans la ShadowMap
 */
void ShadowMap::disable()
{
    // désactiver le FBO
    FrameBufferObject::disable();

    // remettre les modes tels qu'ils étaient avant
    glCullFace(m_CullFacePrec);

    // annuler le décalage
    glPolygonOffset(0.0, 0.0);
    glDisable(GL_POLYGON_OFFSET_FILL);
}


/**
 * cette fonction associe cette shadowmap en tant que texture2D à une unité de texture pour un shader
 * NB: le shader concerné doit être actif
 * @param unit : unité de texture concernée, par exemple GL_TEXTURE0
 * @param locSampler : emplacement de la variable uniform sampler* de cette texture dans le shader ou -1 pour désactiver la texture
 */
void ShadowMap::setTextureUnit(GLint unit, GLint locSampler)
{
    /*****DEBUG*****/
    if (unit < GL_TEXTURE0 || unit > GL_TEXTURE7) {
        throw std::invalid_argument("ShadowMap::setTextureUnit: first parameter, unit is not GL_TEXTURE0 ... GL_TEXTURE7");
    }
    /*****DEBUG*****/

    // activer la shadow map en tant que texture
    glActiveTexture(unit);
    if (locSampler < 0) {
        glBindTexture(GL_TEXTURE_2D, 0);
    } else {
        glBindTexture(GL_TEXTURE_2D, m_DepthBufferId);
        glUniform1i(locSampler, unit-GL_TEXTURE0);
    }
}


/**
 * dessine le depth buffer dans le viewport (pour la mise au point)
 */
void ShadowMap::onDraw()
{
    FrameBufferObject::onDrawDepth();
}


mat4 ShadowMap::c_MatBias;

/** initialise la matrice de biais */
void ShadowMap::staticinit()
{
    c_MatBias = mat4::create();
    mat4::translate(c_MatBias, c_MatBias, vec3::fromValues(0.5, 0.5, 0.5));
    mat4::scale(c_MatBias, c_MatBias, vec3::fromValues(0.5, 0.5, 0.5));
}

File: common/WebGL/Process/ShadowMap.js
//// ce script fournit des fonctions utilitaires pour les programmes
//// du livre Synthèse d'images à l'aide d'OpenGL

// consulter http://blog.tojicode.com/2012/07/using-webgldepthtexture.html

// superclasses et classes nécessaires
Requires("libs/Process/FrameBufferObject.js");


/**
 * C'est un FBO particulier, il n'y a qu'un depth buffer
 * On l'utilise pour dessiner les ombres dans une scène
 */
class ShadowMap extends FrameBufferObject
{
    /**
     * constructeur de la classe ShadowMap
     * @param shadowmapsize : largeur et hauteur de la carte d'ombre
     * @param offsetfill : true si la shadow map doit activer le décalage de polygones
     * @param cullface : mettre gl.FRONT ou gl.BACK pour éliminer les faces avant ou de dos, gl.NONE pour ne rien configurer
     */
    constructor(shadowmapsize, offsetfill=false, cullface=gl.NONE)
    {
        // c'est un FBO ne contenant qu'une depth map de type texture
        super(shadowmapsize, shadowmapsize, gl.TEXTURE_2D, gl.TEXTURE_2D, 0, gl.NEAREST);

        // options de réduction de l'acné
        this.m_OffsetFill = offsetfill;
        this.m_CullFace   = cullface;
    }


    /**
     * supprime cette shadowmap
     */
    destroy()
    {
        super.destroy();
    }


    /**
     * cette méthode change les modes de réduction de l'acné de surface
     * @param offsetfill : true si la shadow map doit activer le décalage de polygones
     * @param cullface : mettre GL_FRONT ou GL_BACK pour éliminer les faces avant ou de dos, GL_NONE pour ne rien configurer
     */
    setAcneReduction(offsetfill, cullface)
    {
        this.m_OffsetFill = offsetfill;
        this.m_CullFace   = cullface;
    }


    /**
     * redirige tous les tracés suivants vers la ShadowMap
     */
    enable()
    {
        // activer le FBO
        super.enable();

        // (optionnel) éliminer les faces avant, afin d'éviter l'acné de surface
        this.m_CullFacePrec = gl.getParameter(gl.CULL_FACE_MODE);
        if (this.m_CullFace != gl.NONE) {
            gl.enable(gl.CULL_FACE);
            gl.cullFace(this.m_CullFace);
        }

        // (optionnel) décalage de polygones, afin d'éviter l'acné de surface
        if (this.m_OffsetFill) {
            gl.enable(gl.POLYGON_OFFSET_FILL);
            gl.polygonOffset(1.0, 1.0);
        }

        // effacer le depth buffer
        gl.clear(gl.DEPTH_BUFFER_BIT);
    }


    /**
     * cesser de rediriger les dessins dans la ShadowMap
     */
    disable()
    {
        // désactiver le FBO
        super.disable();

        // remettre les modes tels qu'ils étaient avant
        gl.cullFace(this.m_CullFacePrec);

        // annuler le décalage
        gl.polygonOffset(0.0, 0.0);
        gl.disable(gl.POLYGON_OFFSET_FILL);
    }


    /**
     * cette fonction associe cette shadowmap en tant que texture2D à une unité de texture pour un shader
     * NB: le shader concerné doit être actif
     * @param unit : unité de texture concernée, par exemple gl.TEXTURE0
     * @param locSampler : emplacement de la variable uniform sampler* de cette texture dans le shader ou null pour désactiver la texture
     */
    setTextureUnit(unit, locSampler=null)
    {
        /*****DEBUG*****/
        if (unit < gl.TEXTURE0 || unit > gl.TEXTURE7) {
            error("ShadowMap.setTextureUnit: first parameter, unit is not gl.TEXTURE0 ... gl.TEXTURE7");
        }
        if (locSampler != null && locSampler.constructor !== WebGLUniformLocation) {
            error("ShadowMap.setTextureUnit: third parameter, locSampler is not either null or the location of a uniform variable : "+locSampler);
        }
        /*****DEBUG*****/

        // activer la shadow map en tant que texture
        gl.activeTexture(unit);
        if (locSampler == null) {
            gl.bindTexture(gl.TEXTURE_2D, null);
        } else {
            gl.bindTexture(gl.TEXTURE_2D, this.m_DepthBufferId);
            gl.uniform1i(locSampler, unit-gl.TEXTURE0);
        }
    }


    /**
     * dessine le depth buffer dans le viewport (pour la mise au point)
     */
    onDraw()
    {
        super.onDrawDepth();
    }
}


/** initialise la matrice de biais */
ShadowMap.c_MatBias = mat4.create();
mat4.translate(ShadowMap.c_MatBias, ShadowMap.c_MatBias, vec3.fromValues(0.5, 0.5, 0.5));
mat4.scale(ShadowMap.c_MatBias, ShadowMap.c_MatBias, vec3.fromValues(0.5, 0.5, 0.5));
